src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -0,0 +1,149 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+use fixedbitset::FixedBitSet;
+use itertools::Itertools;
+use risingwave_common::catalog::Schema;
+use risingwave_common::error::{ErrorCode, Result, RwError};
+use risingwave_pb::plan_common::JoinType;
+
+use super::{
+    ColPrunable, CollectInputRef, LogicalProject, PlanBase, PlanRef, PlanTreeNodeBinary,
+    StreamHashJoin, ToBatch, ToStream,
+};
+use crate::expr::ExprImpl;
+use crate::optimizer::plan_node::batch_nested_loop_join::BatchNestedLoopJoin;
+use crate::optimizer::plan_node::{
+    BatchFilter, BatchHashJoin, EqJoinPredicate, LogicalFilter, LogicalJoin, PlanTreeNode,
+    StreamFilter,
+};
+use crate::optimizer::property::{Distribution, Order};
+use crate::utils::{ColIndexMapping, Condition};
+
+/// `LogicalMultiJoin` combines two or more relations according to some condition.
+///
+/// Each output row has fields from one the inputs. The set of output rows is a subset
+/// of the cartesian product of all the inputs; The `LogicalMultiInnerJoin` is only supported
+/// for inner joins as it implicitly assumes commutativity. Non-inner joins should be
+/// expressed as 2-way `LogicalJoin`s.
+#[derive(Debug, Clone)]
+pub struct LogicalMultiJoin {
+    pub base: PlanBase,
+    inputs: Vec<PlanRef>,
+    on: Condition,
+}
+
+impl fmt::Display for LogicalMultiJoin {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LogicalMultiJoin {{ on: {} }}", &self.on)
+    }
+}
+
+impl LogicalMultiJoin {
+    pub(crate) fn new(base: PlanBase, inputs: Vec<PlanRef>, on: Condition) -> Self {
+        Self { base, inputs, on }
+    }
+
+    pub(crate) fn from_join(join: &PlanRef) -> Option<Self> {
+        let logical_join = join.as_logical_join()?;
+        if logical_join.join_type() != JoinType::Inner {
+            return None;
+        }
+        let left = logical_join.left();
+        let right = logical_join.right();
+
+        let left_col_num = left.schema().len();
+        let right_col_num = right.schema().len();
+
+        let mut inputs = vec![];
+        let mut conjunctions = logical_join.on().conjunctions.clone();
+        if let Some(multi_join) = left.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            conjunctions.extend(multi_join.on().clone());
+        } else {
+            inputs.push(left.clone());
+        }
+        if let Some(multi_join) = right.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            let right_on = multi_join.on().clone();
+            let mut mapping = ColIndexMapping::with_shift_offset(
```

Comment:
Could you add some comments about this mapping?

I thought for the `right_on`, the `input ref` in it should get shifted by `left_col_num` to the right?
`(0...right_col_num) -> (left_col_num...left_col_num + right_col_num)`

Reply:
Yes, basically. But here we are taking the inverse of the o2r mapping, I can change it to yours to be more clear.